## Verdict: NEEDS CHANGES

Overall a clean refactor that consolidates duplicated logging into a single module. The acceptance criteria from #256 are met (timestamps, single source of truth, no remaining inline print calls, tests updated, color detection preserved). Two issues should be addressed before merge:

### 1. Missing unit tests for `pipeline/lib/log.py` (blocking)

This module is the central new code — imported by 7 files — yet has zero direct test coverage. A minimal `test_log.py` (~30 lines) should verify:
- `info()`, `ok()`, `warn()` write to stdout
- `err()` writes to stderr
- Output contains an `HH:MM:SS` timestamp prefix
- ANSI codes are suppressed when `isatty()` is False

Without this, a future change to routing behavior (e.g., reverting `warn()` to stderr) would have no test to flag it.

### 2. Color detection mismatch for `err()` (non-blocking but worth fixing)

`_c()` checks `sys.stdout.isatty()` but `err()` writes to `sys.stderr`. When stdout is piped to a file but stderr is still a TTY (common: `deploy.py run > log.txt`), error messages lose color even though their destination supports it. A simple fix:

```python
def err(msg: str) -> None:
    _ce = lambda code, text: f"\033[{code}m{text}\033[0m" if sys.stderr.isatty() else text
    print(f"{_ts()} {_ce('31', '[ERROR] ')}{msg}", file=sys.stderr)
```

Or pass the stream to `_c()`.

### Notes (non-blocking)

- The `warn()` to stdout standardization is a reasonable design choice given the prior inconsistency (deploy.py used stderr, run.py/prepare.py used stdout). The PR documents this clearly.
- Pre-existing: `get_pods()` in `health.py` silently returns `[]` on kubectl failure without logging — unlike `get_all_pods()` which now properly warns. Not introduced by this PR, but worth a follow-up.

--- a/pipeline/lib/log.py
+++ b/pipeline/lib/log.py
@@ -3,8 +3,9 @@ import sys
 import time
 
 
-def _c(code: str, text: str) -> str:
-    return f"\033[{code}m{text}\033[0m" if sys.stdout.isatty() else text
+def _c(code: str, text: str, *, stream=None) -> str:
+    target = stream or sys.stdout
+    return f"\033[{code}m{text}\033[0m" if target.isatty() else text
 
 
 def _ts() -> str:
@@ -14,4 +15,4 @@ def _ts() -> str:
 def info(msg: str) -> None: print(f"{_ts()} {_c('34', '[INFO]  ')}{msg}")
 def ok(msg: str)   -> None: print(f"{_ts()} {_c('32', '[OK]    ')}{msg}")
 def warn(msg: str) -> None: print(f"{_ts()} {_c('33', '[WARN]  ')}{msg}")
-def err(msg: str)  -> None: print(f"{_ts()} {_c('31', '[ERROR] ')}{msg}", file=sys.stderr)
+def err(msg: str)  -> None: print(f"{_ts()} {_c('31', '[ERROR] ', stream=sys.stderr)}{msg}", file=sys.stderr)

--- a/pipeline/tests/test_log.py
+++ b/pipeline/tests/test_log.py
@@ -1,0 +1,50 @@
+"""Tests for pipeline/lib/log.py."""
+import re
+import sys
+
+from pipeline.lib.log import info, ok, warn, err
+
+
+_TS_PATTERN = re.compile(r"^\d{2}:\d{2}:\d{2} ")
+
+
+def test_info_writes_to_stdout(capsys):
+    info("hello")
+    out = capsys.readouterr()
+    assert "hello" in out.out
+    assert out.err == ""
+
+
+def test_ok_writes_to_stdout(capsys):
+    ok("done")
+    out = capsys.readouterr()
+    assert "done" in out.out
+    assert out.err == ""
+
+
+def test_warn_writes_to_stdout(capsys):
+    warn("careful")
+    out = capsys.readouterr()
+    assert "careful" in out.out
+    assert out.err == ""
+
+
+def test_err_writes_to_stderr(capsys):
+    err("broken")
+    out = capsys.readouterr()
+    assert "broken" in out.err
+    assert "broken" not in out.out
+
+
+def test_timestamp_prefix(capsys):
+    info("test msg")
+    out = capsys.readouterr().out
+    assert _TS_PATTERN.match(out)
+
+
+def test_no_ansi_when_not_tty(capsys, monkeypatch):
+    monkeypatch.setattr(sys.stdout, "isatty", lambda: False)
+    info("plain")
+    out = capsys.readouterr().out
+    assert "\033[" not in out
+    assert "plain" in out

